src/realm/sync/noinst/client_reset.cpp

```diff
@@ -981,41 +987,87 @@ LocalVersionIDs perform_client_reset_diff(DBRef db_local, DBRef db_remote, sync:
-        REALM_ASSERT_RELEASE(client_repl);
-        ChangesetEncoder& encoder = client_repl->get_instruction_encoder();
-        const sync::ChangesetEncoder::Buffer& buffer = encoder.buffer();
-        recovered_changeset = {buffer.data(), buffer.size()};
+    // FLX with recovery has to be done in multiple commits, which is significantly different than other modes
+    if (recover_local_changes && sub_store) {
+        // In FLX recovery, save a copy of the pending subscriptions for later. This
+        // needs to be done before they are wiped out by remake_active_subscription()
+        std::vector<SubscriptionSet> pending_subscriptions = sub_store->get_pending_subscriptions();
+        // transform the local Realm such that all public tables become identical to the remote Realm
+        transfer_group(*wt_remote, *wt_local, logger);
+        // now that the state of the fresh and local Realms are identical,
+        // reset the local sync history.
+        // Note that we do not set the new file ident yet! This is done in the last commit.
+        history_local->set_client_reset_adjustments(current_version_local, orig_file_ident, fresh_server_version,
+                                                    recovered_changeset);
+        // The local Realm is committed. There are no changes to the remote Realm.
+        wt_remote->rollback_and_continue_as_read();
+        wt_local->commit_and_continue_as_read();
+        // Make a copy of the active subscription set and mark it as
+        // complete. This will cause all other subscription sets to become superceded.
+        remake_active_subscription();
+        // Apply local changes interleaved with pending subscriptions in separate commits
+        // as needed. This has the consequence that there may be extra notifications along
+        // the way to the final state, but since separate commits are necessary, this is
+        // unavoidable.
+        wt_local = db_local->start_write();
+        RecoverLocalChangesetsHandler handler{*wt_local, *frozen_pre_local_state, logger};
+        handler.process_changesets(local_changes, std::move(pending_subscriptions)); // throws on error
+        // The new file ident is set as part of the final commit. This is to ensure that if
+        // there are any exceptions during recovery, or the process is killed for some
+        // reason, the client reset cycle detection will catch this and we will not attempt
+        // to recover again. If we had set the ident in the first commit, a Realm which was
+        // partially recovered, but interrupted may continue sync the next time it is
+        // opened with only partially recovered state while having lost the history of any
+        // offline modifications.
+        history_local->set_client_file_ident_in_wt(current_version_local, client_file_ident);
+        wt_local->commit_and_continue_as_read();
     }
+    else {
+        if (recover_local_changes) {
+            REALM_ASSERT(!sub_store);
```

Comment:
this is not really needed. if you end up in this code path then `sub_store` does not exist.